### PR TITLE
CodAn and MITGARD recipes

### PR DESCRIPTION
**Pull request by pedronachtigall**
Describe your pull request here

----

Please read the [guidelines for Bioconda recipes](https://bioconda.github.io/contributor/guidelines.html) before opening a pull request (PR).

* If this PR adds or updates a recipe, use "Add" or "Update" appropriately as the first word in its title.
* New recipes not directly relevant to the biological sciences need to be submitted to the [conda-forge channel](https://conda-forge.org/docs/) instead of Bioconda.
* PRs require reviews prior to being merged. Once your PR is passing tests and ready to be merged, please issue the `@BiocondaBot please add label` command.
* Please post questions [on Gitter](https://gitter.im/bioconda/Lobby) or ping `@bioconda/core` in a comment.

<details>
  <summary>Please use the following BiocondaBot commands:</summary>

Everyone has access to the following BiocondaBot commands, which can be given in a comment:

<table>
  <tr>
    <td><code>@BiocondaBot please update</code></td>
    <td>Merge the master branch into a PR.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please add label</code></td>
    <td>Add the <code>please review & merge</code> label.</td>
  </tr>
  <tr>
    <td><code>@BiocondaBot please fetch artifacts</code></td>
    <td>Post links to CI-built packages/containers. <br />You can use this to test packages locally.</td>
  </tr>
</table>

For members of the Bioconda project, the following command is also available:

<table>
  <tr>
    <td><code>@BiocondaBot please merge</code></td>
    <td>Upload built packages/containers and merge a PR. <br />Someone must approve a PR first! <br />This reduces CI build time by reusing built artifacts.</td>
  </tr>
</table>

Also, the bot watches for comments from non-members that include `@bioconda/<team>` and will automatically re-post them to notify the addressed `<team>`.

</details>

### CODE REVIEW

**Comment by pedronachtigall** [comment]
Hi @bernt-matthias , here comes the PR to mitgard and I am also doing a PR to another tool named CodAn (https://github.com/pedronachtigall/CodAn). Hoe it works fine for both tools.

File: recipes/codan/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
+{% set version = "1.0" %}
+
+package:
+  name: CodAn
```
Comment (bernt-matthias) [inline recipes/codan/meta.yaml:4]: I think name needs to be lower case

**Review by bernt-matthias — COMMENTED** [review]
You can find  more problems by looking at the failing CI checks.

File: recipes/codan/meta.yaml (hunk 1 above)
Comment (pedronachtigall) [reply recipes/codan/meta.yaml:4]: Hi @bernt-matthias. Thank you for your reply. I will change it and re-send the PR.

**Comment by pedronachtigall** [comment]
@BiocondaBot please update
@BiocondaBot please add label
@BiocondaBot please fetch artifacts

**Comment by pedronachtigall** [comment]
@BiocondaBot please review & merge

**Comment by bernt-matthias** [comment]
Hi @pedronachtigall please check the output of the CI system.. You find it at the bottom of this page. Currently Linting is failing with the following messages

```
ERROR: recipes/codan/meta.yaml:1: missing_build_number: The recipe is missing a build number
ERROR: recipes/codan/meta.yaml:9: missing_hash: The recipe is missing a checksum for a source file
ERROR: recipes/codan/meta.yaml:1: should_be_noarch_generic: The recipe should be build as `noarch`
ERROR: recipes/codan/meta.yaml:8: uses_vcs_url: The recipe downloads source from a VCS
WARNING: recipes/codan/meta.yaml:32: gpl_requires_license_distributed: The recipe packages GPL software but is missing copy of license.
ERROR: recipes/mitgard/meta.yaml:1: missing_build_number: The recipe is missing a build number
ERROR: recipes/mitgard/meta.yaml:9: missing_hash: The recipe is missing a checksum for a source file
ERROR: recipes/mitgard/meta.yaml:1: should_be_noarch_generic: The recipe should be build as `noarch`
ERROR: recipes/mitgard/meta.yaml:8: uses_vcs_url: The recipe downloads source from a VCS
WARNING: recipes/mitgard/meta.yaml:33: gpl_requires_license_distributed: The recipe packages GPL software but is missing copy of license.
Errors were found
```

**Review by pedronachtigall — COMMENTED** [review]
@BiocondaBot please review & merge

**Comment by bernt-matthias** [comment]
Still two linter problems: 

```
ERROR: recipes/codan/meta.yaml:1: missing_build_number: The recipe is missing a build number
ERROR: recipes/codan/meta.yaml:1: should_be_noarch_generic: The recipe should be build as `noarch`
ERROR: recipes/mitgard/meta.yaml:1: missing_build_number: The recipe is missing a build number
ERROR: recipes/mitgard/meta.yaml:1: should_be_noarch_generic: The recipe should be build as `noarch`
```

Note that the bioconda bot messages do not make much sense until all checks are successful.

File: recipes/codan/meta.yaml [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+package:
+  name: codan
+  version: {{ version }}
+
+source:
+  url: https://github.com/pedronachtigall/CodAn/archive/refs/tags/v{{ version }}.tar.gz
+  sha256: 561d212b5acf8201dfa1af51688a6d9b0e7b8d97c1c328571987b4dda6369147
+
+requirements:
+  build:
```
Comment (bernt-matthias) [inline recipes/codan/meta.yaml:12]: This (the build block) should be moved toplevel. There can be a build section under requirements, but this should list build requirements.

Comment (pedronachtigall) [reply recipes/codan/meta.yaml]: @bernt-matthias now the lint passed, but we are having problems in the linux-test. How can I check the output of CI system without logging into CI?

File: recipes/codan/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+  url: https://github.com/pedronachtigall/CodAn/archive/refs/tags/v{{ version }}.tar.gz
+  sha256: 561d212b5acf8201dfa1af51688a6d9b0e7b8d97c1c328571987b4dda6369147
+
+build:
+  noarch: generic
+  number: 0
+
+requirements:
+  host:
+    - python=3.7
```
Comment (bernt-matthias) [inline recipes/codan/meta.yaml:17]: Can you make this >=16?

**Review by bernt-matthias — COMMENTED** [review]
Mitgard is kind of successful. Seems that one of the requirements needs bowie 1.x.. Not sure how to solve this.

File: recipes/codan/meta.yaml (hunk 2 above)
Comment (bernt-matthias) [reply recipes/codan/meta.yaml]: You need to login. But there should be a button "login with github" which simplifies this.

**Comment by pedronachtigall** [comment]
> Mitgard is kind of successful. Seems that one of the requirements needs bowie 1.x.. Not sure how to solve this.

Hi @bernt-matthias ,
Trinity uses bowtie 1.x (https://bioconda.github.io/recipes/bowtie/README.html), but it is available to be installed through conda.
Should I added it as a requirement in the meta.yaml file?

Thanks for your help.

**Review by bernt-matthias — COMMENTED** [review]
codan fails because: `cp: cannot stat ‘bin/*’: No such file or directory`

**Comment by bernt-matthias** [comment]
Seems that the release misses the `bin` folder

**Comment by pedronachtigall** [comment]
> Seems that the release misses the `bin` folder

Yep, you are right. At the time I just had a tarball in the github repository. I just release version 1.1 to solve this issue.
Thank you for the help.

**Comment by bernt-matthias** [comment]
Looks good. Can you test the packages provided as artifacts with a real world example?

**Comment by pedronachtigall** [comment]
> Looks good. Can you test the packages provided as artifacts with a real world example?

@bernt-matthias 

MITGARD is ok on all computers I tested. It worked like a charm all the time.

CodAn have some issues with a few of the times I tested. When it has issues it was related to the fact that codan messes with Perl libraries path (see https://stackoverflow.com/a/62141283). For instance, it installs Perl version 5.26 but installs all libraries into path 5.22, which makes the Perl not find any of the installed libraries (in the case bioperl and MCE). To solve this I had to indicate the path with export with something like "export PERL5LIB=/path/to/miniconda3/envs/codan_env/lib/perl5/site_perl/5.22.0/".

What is the best thing to do (in your opinion)? it will be better to try to direct the meta.yaml to install Perl version 5.22 to not have such issues, or let it as this current version but add and warning related to this issue that it can happen but can be solved by using the "export" as above.

Thank you for the help.

**Comment by bernt-matthias** [comment]
I'm absolutely no perl expert. Could you ask at https://gitter.im/bioconda/Lobby ? Maybe also ask if the path mangling due to bowtie and bowtie2 needs attention...

**Comment by pedronachtigall** [comment]
After reading and checking about this issue, I noticed that it is a conda bug (related to the latest conda version). To solve the problem the user must set the PERL5LIB PATH by using `export PERL5LIB=$CONDA_PREFIX/lib/perl5/site_perl/5.22.0/` all the time it uses the environment or by using `conda env config vars set PERL5LIB=$CONDA_PREFIX/lib/perl5/site_perl/5.22.0/ -n [env_name]` to make it permanent for that environment.
I report the bug to the conda team and hope they work on it to have it fixed in further conda releases.

I think we can move by adding both packages into bioconda. For now, I will keep a warning about this issue in the README of CodAn to save user's time. When conda team fixes it, we will be able to have it properly running after any new installation.

**Comment by bernt-matthias** [comment]
There seems to be the possibility to automate setting the path: https://docs.conda.io/projects/conda-build/en/latest/resources/activate-scripts.html

**Comment by bernt-matthias** [comment]
@BiocondaBot please add label

**Review by dpryan79 — APPROVED** [review]

**Comment by dpryan79** [comment]
@bioconda-bot please merge

**Pull request merged**